### ISSUE TITLE
Rank Expansion & Training

### ISSUE THREAD

**Issue by steve2972** [clovaai/rexnet#11]
While programming a Tensorflow implementation of ReXNet, I had a couple of questions regarding your proposed rank expansion and training methodologies.

**First Question**

First of all, regarding your implementation of the [Linear Bottleneck] residual connection, the code says it aggregates the input into the first layers of the bottleneck with the dimensions equal to the input_channels already inside the sequential output.
``` python
if self.use_shortcut:
    out[:, 0:self.in_channels] += x
```
However, when reading the ResNet paper, the authors of the residual block said that when the input channels and output channels are different, it is necessary for x to be linearly projected to the same shape and then added to the residual function. My implementation of this can be seen as follows:
``` python
# Tensorflow implementation of ReXNet
# y = the linear bottleneck model (ConvBNSwish + ConvAct + Squeeze + etc.)
if use_shortcut:
    x = Conv2D(filters=self.out_channels, strides=self.stride,...)(_input)
    x = BatchNormalization(...)(x)
    y += x
return y
```

I was wondering if there was a particular reason why you did not opt to use a linear projection on the input, and if there are any significant performance differences between the two methodologies.

**Second Question**

Furthermore, I noticed in the [Linear Bottleneck] code that you only provided a depthwise operation without a pointwise convolution. The code provided says
``` python
ConvBNAct(out, in_channels=dw_channels, channels=dw_channels, kernel=3, stride=stride, pad=1,
                  num_group=dw_channels, active=False)
```
Does this mean this is **not** a depthwise separable convolution? Or is this also a method to reduce the representational bottleneck?

Thank you!

<br>

---

안녕하세요,

좋은 논문과 그에 대한 코드를 제공해주셔서 감사합니다.  Mobilenet에 이어서 경량화 뉴럴넷을 찾아보는 중 ReXNet을 발견하고 좋은 성능을 보여줬기에 관심이 생겼습니다. 그래서 실제로 모바일 핸드폰에 돌리기 위해 Tensorflow로 구현해보면서 몇 가지 질문들이 생겨서 이슈를 남깁니다.

감사합니다!

**Comment by dyhan0920** [comment]
Hello @steve2972! Sorry for the late reply.

For the first question, the implementation using simple addition is to take advantage of a simple implementation (in Pytorch) of shortcuts without any additional parameters. I have not tested the shortcut connection with a projection using 1x1 conv with BN but guess using projection shortcut may have a possibility to improve the accuracy due to extra parameters. Notice that the inverted bottleneck with the stride of 2 does not have a shortcut.

Next, after the depthwise convolution, a pointwise convolution follows: https://github.com/clovaai/rexnet/blob/104f2184754d97b8050f24bdf39f650b3e80cad8/rexnetv1.py#L112. 

Best regards,
Dongyoon

모델에 관심을 가져주셔서 감사합니다!  여기 issue나 제 개인 메일을 통해서 언제든지 궁금하신 사항 남겨주시면 답변 드리겠습니다.

감사합니다.

**Comment by bonlime** [comment]
Regarding your first question - while authors probably didn't experiment with this design choice, there is a paper which proposes to use partial residuals and shows that it has some nice properties in terms of gradient propagation - [Partial Residual Networks](https://github.com/WongKinYiu/PartialResidualNetworks)

**Comment by dyhan0920** [comment]
@bonlime Thanks for the suggestion. Our model is an instance model to provide see how the rank expansion affects the model performance, so one can replace the basic modules with any components. I will put the partial residual module into the list of things to do in the future.
